cpp/src/join/hash_join.cu

```diff
@@ -423,58 +475,53 @@ std::size_t hash_join<Hasher>::full_join_size(cudf::table_view const& probe,
   // Trivial left join case - exit early
   if (_is_empty) { return probe.num_rows(); }
 
-  auto flattened_probe =
-    structs::detail::flatten_nested_columns(probe,
-                                            {},
-                                            {},
-                                            structs::detail::column_nullability::FORCE,
-                                            stream,
-                                            rmm::mr::get_current_device_resource());
-  auto const flattened_probe_table = flattened_probe->flattened_columns();
-
-  auto build_table_ptr           = cudf::table_device_view::create(_build, stream);
-  auto flattened_probe_table_ptr = cudf::table_device_view::create(flattened_probe_table, stream);
+  auto preprocessed_probe =
+    cudf::experimental::row::equality::preprocessed_table::create(probe, stream);
 
   return cudf::detail::get_full_join_size(
-    *build_table_ptr,
-    *flattened_probe_table_ptr,
+    _build,
+    probe,
+    _preprocessed_build,
+    preprocessed_probe,
     _hash_table,
-    cudf::has_nulls(flattened_probe_table) | cudf::has_nulls(_build),
+    cudf::has_nested_nulls(probe) | cudf::has_nested_nulls(_build),
     _nulls_equal,
     stream,
     mr);
 }
 
 template <typename Hasher>
-template <cudf::detail::join_kind JoinKind>
 std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
           std::unique_ptr<rmm::device_uvector<size_type>>>
 hash_join<Hasher>::probe_join_indices(cudf::table_view const& probe_table,
+                                      cudf::detail::join_kind join,
                                       std::optional<std::size_t> output_size,
                                       rmm::cuda_stream_view stream,
                                       rmm::mr::device_memory_resource* mr) const
 {
   // Trivial left join case - exit early
-  if (_is_empty and JoinKind != cudf::detail::join_kind::INNER_JOIN) {
+  if (_is_empty and join != cudf::detail::join_kind::INNER_JOIN) {
     return get_trivial_left_join_indices(probe_table, stream, mr);
   }
 
   CUDF_EXPECTS(!_is_empty, "Hash table of hash join is null.");
 
-  auto build_table_ptr = cudf::table_device_view::create(_build, stream);
-  auto probe_table_ptr = cudf::table_device_view::create(probe_table, stream);
-
-  auto join_indices = cudf::detail::probe_join_hash_table<JoinKind>(
-    *build_table_ptr,
-    *probe_table_ptr,
+  auto preprocessed_probe =
+    cudf::experimental::row::equality::preprocessed_table::create(probe_table, stream);
+  auto join_indices = cudf::detail::probe_join_hash_table(
```

Comment:
`const`.

Reply:
`join_indices` cannot be `const` because being called into `detail::concatenate_vector_pairs` which does not accept `const`